pkg/validation/validation_test.go

```diff
@@ -155,6 +156,185 @@ func TestContainerPorts(t *testing.T) {
 	}
 }
 
+func TestLoadBalancerIP(t *testing.T) {
+	testCases := []struct {
+		description string
+		address     string
+		expected    bool
+	}{
+		{
+			description: "default load balancer type service without IP specified",
+			expected:    true,
+		},
+		{
+			description: "user-specified load balancer IPv4 address",
+			address:     "1.2.3.4",
+			expected:    true,
+		},
+		{
+			description: "user-specified load balancer IPv6 address",
+			address:     "2607:f0d0:1002:51::4",
+			expected:    true,
+		},
+		{
+			description: "invalid IPv4 address",
+			address:     "1.2..4",
+			expected:    false,
+		},
+		{
+			description: "invalid IPv6 address",
+			address:     "2607:f0d0:1002:51:::4",
+			expected:    false,
+		},
+	}
+
+	name := "test-validation"
+	cntr := &operatorv1alpha1.Contour{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: fmt.Sprintf("%s-ns", name),
+		},
+		Spec: operatorv1alpha1.ContourSpec{
+			Namespace: operatorv1alpha1.NamespaceSpec{Name: "projectcontour"},
+			NetworkPublishing: operatorv1alpha1.NetworkPublishing{
+				Envoy: operatorv1alpha1.EnvoyNetworkPublishing{
+					Type: operatorv1alpha1.LoadBalancerServicePublishingType,
+					LoadBalancer: operatorv1alpha1.LoadBalancerStrategy{
+						Scope: "External",
+						ProviderParameters: operatorv1alpha1.ProviderLoadBalancerParameters{
+							Type: "GCP",
+							GCP:  &operatorv1alpha1.GCPLoadBalancerParameters{},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		if tc.address != "" {
+			cntr.Spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.GCP.Address = &tc.address
+		}
+
+		err := validation.LoadBalancerAddress(cntr)
+		if err != nil && tc.expected {
+			t.Fatalf("%q: failed with error: %#v", tc.description, err)
+		}
+		if err == nil && !tc.expected {
+			t.Fatalf("%q: expected to fail but received no error", tc.description)
+		}
+	}
+}
+
+func TestLoadBalancerProvider(t *testing.T) {
+	testCases := []struct {
+		description        string
+		provider           string
+		additionalProvider string
```

Comment:
I suggest using [LoadBalancerProviderType](https://github.com/projectcontour/contour-operator/blob/release-1.16/api/v1alpha1/contour_types.go#L286-L291) as the value instead of string for these 2 fields.